File: v1/tools/can_timing.py
#!/usr/bin/env python3
"""CAN bit timing, computed independently of the firmware's C solver.

This exists to be a second opinion.  can_timing.c runs on the target and picks
the register values the bus actually sees; if it is wrong, every frame is
wrong, and the symptom -- occasional errors that get worse with cable length
-- is indistinguishable from a dozen other problems.  So the numbers get
computed twice, by two implementations written to the same specification
rather than one ported from the other, and tests/test_can_timing_crosscheck.py
asserts they agree.

Usage:
    python3 can_timing.py --clock 170000000 --bitrate 500000
    python3 can_timing.py --clock 170000000 --bitrate 2000000 --fd
    python3 can_timing.py --table
"""

from __future__ import annotations

import argparse
from dataclasses import dataclass

# STM32G4 FDCAN register limits.  The data phase registers are much smaller
# than the nominal ones, and a solver that ignores that produces values which
# silently truncate when written.
NBRP_MAX = 512
NTSEG1_MAX = 256
NTSEG2_MAX = 128
NSJW_MAX = 128

DBRP_MAX = 32
DTSEG1_MAX = 32
DTSEG2_MAX = 16
DSJW_MAX = 16

# CiA 601-2's default, and the one to use when nothing else is specified.
DEFAULT_SAMPLE_POINT_PERMIL = 875

# Time quanta beyond this buy nothing.  Below it, resolution is the binding
# constraint: a five-tq bit can express a sample point exactly, but it leaves
# SJW at 1 and no room for transmitter delay compensation.  Ranking resolution
# above sample point up to this floor -- and ignoring it above -- is what stops
# the solver answering 2 Mbit/s CAN FD with five time quanta.
RESOLUTION_ENOUGH_TQ = 16

OSC_PPM_HSI16 = 10000
OSC_PPM_CRYSTAL_30 = 30
OSC_PPM_CRYSTAL_50 = 50


@dataclass(frozen=True)
class Timing:
    valid: bool = False
    brp: int = 0
    tseg1: int = 0
    tseg2: int = 0
    sjw: int = 0
    nominal_bt_tq: int = 0
    bitrate_bps: int = 0
    bitrate_error_ppm: int = 0
    sample_point_permil: int = 0

    @property
    def register_nbtp(self) -> int:
        return (((self.sjw - 1) & 0x7F) << 25 | ((self.brp - 1) & 0x1FF) << 16
                | ((self.tseg1 - 1) & 0xFF) << 8 | ((self.tseg2 - 1) & 0x7F))

    @property
    def register_dbtp(self) -> int:
        return (((self.brp - 1) & 0x1F) << 16 | ((self.tseg1 - 1) & 0x1F) << 8
                | ((self.tseg2 - 1) & 0xF) << 4 | ((self.sjw - 1) & 0xF))


def max_tolerance_ppm(t: Timing) -> int:
    """The oscillator tolerance this bit timing can absorb, per node.

    ISO 11898-1 gives two bounds and the answer is the smaller:

        df <= sjw / (20 * nbt)
        df <= min(phase_seg1, phase_seg2) / (2 * (13 * nbt - phase_seg2))

    The first is resynchronisation headroom; the second is the five-bit
    stretch a receiver has to tolerate between resynchronisation
    opportunities.  Which one binds depends on the sample point, which is why
    both have to be evaluated rather than quoting the familiar one.
    """
    if not t.valid or t.nominal_bt_tq == 0:
        return 0

    nbt = t.nominal_bt_tq
    ps2 = t.tseg2
    ps1 = t.tseg1  # prop_seg + phase_seg1; the bound uses the pair together

    term_sjw = (t.sjw * 1_000_000) // (20 * nbt)

    smaller = min(ps1, ps2)
    denom = 2 * (13 * nbt - ps2)
    term_phase = (smaller * 1_000_000) // denom if denom > 0 else 0

    return min(term_sjw, term_phase)


def osc_ok(t: Timing, osc_ppm: int) -> bool:
    """Both nodes contribute error, so each must fit inside the budget."""
    return max_tolerance_ppm(t) >= osc_ppm


def score(t: Timing, want_sp: int) -> tuple:
    """The preference order, as a sort key -- lower is better.

    This is the specification the two implementations share.  How each one
    searches is deliberately different, so that a disagreement means
    something; what counts as a better answer has to be the same, or the
    comparison is meaningless.

      1. Bit rate error, because a bit rate that is merely close is a bus that
         works until it does not.
      2. Resolution, up to RESOLUTION_ENOUGH_TQ and no further.  Below the
         floor, more time quanta genuinely matter; above it they are free and
         should not outrank anything.
      3. Sample point, exactly as requested.  An earlier sample point is a
         real change in how the bus behaves on a long cable, so this is not
         somewhere to be approximate -- an earlier version of this solver had
         a five-percent tie band here and quietly returned 82.7% when asked
         for 87.5%, because the wider phase segment 2 bought it a better
         tolerance score.  That is a trade a person should make deliberately
         by asking for a different sample point, not one a solver should make
         on their behalf.
      4. Oscillator tolerance, because among genuine equals, prefer the timing
         that survives a worse crystal.
      5. More time quanta, purely so the answer is deterministic.
    """
    return (t.bitrate_error_ppm,
            -min(t.nominal_bt_tq, RESOLUTION_ENOUGH_TQ),
            abs(t.sample_point_permil - want_sp),
            -max_tolerance_ppm(t),
            -t.nominal_bt_tq)


def solve(clock_hz: int, bitrate_bps: int,
          sample_point_permil: int = DEFAULT_SAMPLE_POINT_PERMIL,
          data_phase: bool = False) -> Timing:
    """Exhaustive search over every legal (brp, tseg1, tseg2).

    The C solver does a local search: it computes the ideal tseg1 for each
    prescaler and examines that value plus its two neighbours.  That is fast
    and fits on the target, but a local search can miss.  This one enumerates
    the whole space, which is slow and cannot, so the cross-check can tell the
    difference between "the two agree" and "the two agree because one copied
    the other".
    """
    if clock_hz == 0 or bitrate_bps == 0:
        return Timing()
    if sample_point_permil <= 0 or sample_point_permil >= 1000:
        sample_point_permil = DEFAULT_SAMPLE_POINT_PERMIL

    brp_max = DBRP_MAX if data_phase else NBRP_MAX
    tseg1_max = DTSEG1_MAX if data_phase else NTSEG1_MAX
    tseg2_max = DTSEG2_MAX if data_phase else NTSEG2_MAX
    sjw_max = DSJW_MAX if data_phase else NSJW_MAX

    best: Timing | None = None
    best_key: tuple | None = None

    for brp in range(1, brp_max + 1):
        # The hardware divides the kernel clock by brp, then counts nbt time
        # quanta per bit.  Only the total brp * nbt matters to the bit rate,
        # so enumerate nbt directly instead of deriving it.
        for nbt in range(4, 2 + tseg1_max + tseg2_max):
            divisor = brp * nbt
            actual = clock_hz // divisor
            if actual == 0:
                break
            error_ppm = abs(actual - bitrate_bps) * 1_000_000 // bitrate_bps
            # Anything beyond a few percent is never going to win; skipping it
            # keeps the exhaustive search tolerable.
            if error_ppm > 50_000:
                continue

            for tseg1 in range(1, min(tseg1_max, nbt - 2) + 1):
                tseg2 = nbt - 1 - tseg1
                if tseg2 < 1 or tseg2 > tseg2_max:
                    continue

                cand = Timing(
                    valid=True, brp=brp, tseg1=tseg1, tseg2=tseg2,
                    sjw=min(tseg2, sjw_max), nominal_bt_tq=nbt,
                    bitrate_bps=actual, bitrate_error_ppm=error_ppm,
                    sample_point_permil=((1 + tseg1) * 1000) // nbt,
                )
                key = score(cand, sample_point_permil)
                if best_key is None or key < best_key:
                    best, best_key = cand, key

    return best or Timing()


def tdco(t: Timing, transceiver_delay_ns: int, clock_hz: int) -> int:
    """Secondary sample point offset, in FDCAN clock cycles.

    Above roughly 1 Mbit/s the transceiver's loop delay is a real fraction of
    a data bit, and without compensation the controller samples its own
    transmission at the wrong moment and reports bit errors on a bus that is
    working correctly.
    """
    if not t.valid or clock_hz == 0:
        return 0
    delay_cycles = (transceiver_delay_ns * clock_hz) // 1_000_000_000
    ssp = delay_cycles + t.brp * (1 + t.tseg1)
    return min(ssp, 127)


def _report(t: Timing, label: str, clock_hz: int) -> None:
    if not t.valid:
        print(f"{label}: NO SOLUTION")
        return
    tol = max_tolerance_ppm(t)
    print(f"{label}")
    print(f"  bit rate        {t.bitrate_bps:,} bit/s "
          f"(error {t.bitrate_error_ppm} ppm)")
    print(f"  time quanta     {t.nominal_bt_tq} "
          f"(brp {t.brp}, tq = {clock_hz / t.brp / 1e6:.3f} MHz)")
    print(f"  tseg1 / tseg2   {t.tseg1} / {t.tseg2}")
    print(f"  sjw             {t.sjw}")
    print(f"  sample point    {t.sample_point_permil / 10:.1f}%")
    print(f"  osc tolerance   {tol} ppm per node")
    print(f"    30 ppm xtal   {'OK' if osc_ok(t, OSC_PPM_CRYSTAL_30) else 'FAIL'}")
    print(f"    50 ppm xtal   {'OK' if osc_ok(t, OSC_PPM_CRYSTAL_50) else 'FAIL'}")
    print(f"    HSI16 +/-1%   {'OK' if osc_ok(t, OSC_PPM_HSI16) else 'FAIL'}"
          "   <- section 7.2")
    print(f"  NBTP            0x{t.register_nbtp:08X}")
    print(f"  DBTP            0x{t.register_dbtp:08X}")


def main() -> int:
    ap = argparse.ArgumentParser(description=__doc__,
                                 formatter_class=argparse.RawDescriptionHelpFormatter)
    ap.add_argument("--clock", type=int, default=170_000_000,
                    help="FDCAN kernel clock in Hz (default 170 MHz)")
    ap.add_argument("--bitrate", type=int, default=500_000)
    ap.add_argument("--sample-point", type=int,
                    default=DEFAULT_SAMPLE_POINT_PERMIL,
                    help="per mil, e.g. 875 for 87.5%%")
    ap.add_argument("--fd", action="store_true",
                    help="solve within the tighter data-phase register limits")
    ap.add_argument("--table", action="store_true",
                    help="every standard rate, plus the FD data phases")
    args = ap.parse_args()

    if args.table:
        print(f"FDCAN kernel clock: {args.clock:,} Hz\n")
        header = (f"{'rate':>10} {'phase':>8} {'brp':>4} {'tseg1':>6} "
                  f"{'tseg2':>6} {'sjw':>4} {'tq':>4} {'SP':>7} "
                  f"{'err ppm':>8} {'tol ppm':>8}  crystal HSI16")
        print(header)
        print("-" * len(header))
        rows = [(r, False, DEFAULT_SAMPLE_POINT_PERMIL)
                for r in (125_000, 250_000, 500_000, 1_000_000)]
        rows += [(r, True, 800) for r in (2_000_000, 5_000_000)]
        for rate, fd, sp in rows:
            t = solve(args.clock, rate, sp, fd)
            if not t.valid:
                print(f"{rate:>10,} {'data' if fd else 'nominal':>8}  NO SOLUTION")
                continue
            print(f"{rate:>10,} {'data' if fd else 'nominal':>8} {t.brp:>4} "
                  f"{t.tseg1:>6} {t.tseg2:>6} {t.sjw:>4} {t.nominal_bt_tq:>4} "
                  f"{t.sample_point_permil / 10:>6.1f}% "
                  f"{t.bitrate_error_ppm:>8} {max_tolerance_ppm(t):>8}  "
                  f"{'OK   ' if osc_ok(t, OSC_PPM_CRYSTAL_30) else 'FAIL '}   "
                  f"{'OK' if osc_ok(t, OSC_PPM_HSI16) else 'FAIL'}")
        print("\nThe HSI16 column is section 7.2: there is no standard bit "
              "rate at which\nskipping the crystal is acceptable.")
        return 0

    t = solve(args.clock, args.bitrate, args.sample_point, args.fd)
    _report(t, "data phase" if args.fd else "nominal phase", args.clock)
    return 0 if t.valid else 1


if __name__ == "__main__":
    raise SystemExit(main())

File: v1/firmware/core/src/can_timing.c
/* See can_timing.h. */

#include "can_timing.h"

#include <stdlib.h>

static int abs_i(int v) { return v < 0 ? -v : v; }

/* Time quanta beyond this buy nothing, and below it resolution is the binding
 * constraint -- see the ranking comment in can_timing_solve(). */
#define RESOLUTION_ENOUGH_TQ 16

static uint32_t min_tq(uint32_t nbt) {
    return nbt < RESOLUTION_ENOUGH_TQ ? nbt : RESOLUTION_ENOUGH_TQ;
}

static uint32_t min_u32(uint32_t a, uint32_t b) { return a < b ? a : b; }

int can_timing_max_tolerance_ppm(const can_timing_t *t) {
    if (t == NULL || !t->valid || t->nominal_bt_tq == 0) {
        return 0;
    }

    const int64_t nbt = (int64_t)t->nominal_bt_tq;
    const int64_t ps1 = (int64_t)t->tseg1;
    const int64_t ps2 = (int64_t)t->tseg2;
    const int64_t sjw = (int64_t)t->sjw;

    /* df <= SJW / (2 * 10 * NBT) */
    const int64_t term_sjw = (sjw * 1000000) / (20 * nbt);

    /* df <= min(PS1, PS2) / (2 * (13 * NBT - PS2)) */
    const int64_t denom = 2 * (13 * nbt - ps2);
    const int64_t smaller = (ps1 < ps2) ? ps1 : ps2;
    const int64_t term_phase = (denom > 0) ? (smaller * 1000000) / denom : 0;

    const int64_t worst = (term_sjw < term_phase) ? term_sjw : term_phase;
    return (int)worst;
}

bool can_timing_osc_ok(const can_timing_t *t, int osc_ppm) {
    return can_timing_max_tolerance_ppm(t) >= osc_ppm;
}

can_timing_t can_timing_solve(uint32_t clock_hz, uint32_t bitrate_bps,
                              int sample_point_permil, bool data_phase) {
    can_timing_t best = {0};
    best.valid = false;

    if (clock_hz == 0 || bitrate_bps == 0) {
        return best;
    }
    if (sample_point_permil <= 0 || sample_point_permil >= 1000) {
        sample_point_permil = 875; /* CiA's recommendation */
    }

    const uint32_t brp_max = data_phase ? FDCAN_DBRP_MAX : FDCAN_NBRP_MAX;
    const uint32_t tseg1_max = data_phase ? FDCAN_DTSEG1_MAX : FDCAN_NTSEG1_MAX;
    const uint32_t tseg2_max = data_phase ? FDCAN_DTSEG2_MAX : FDCAN_NTSEG2_MAX;
    const uint32_t sjw_max = data_phase ? FDCAN_DSJW_MAX : FDCAN_NSJW_MAX;

    int best_rate_err = 0;
    int best_sp_err = 0;
    int best_tol = 0;
    uint32_t best_res = 0;

    for (uint32_t brp = 1; brp <= brp_max; brp++) {
        const uint32_t floor_tq = clock_hz / (brp * bitrate_bps);
        if (floor_tq < 3) {
            break; /* prescaler is already too large; more will not help */
        }

        /* Both roundings of the time quanta count, not just the one that
         * truncating division happens to give.
         *
         * When the bit rate divides the kernel clock exactly there is only
         * one candidate and the second iteration is wasted work.  When it
         * does not -- 800 kbit/s at 170 MHz wants 212.5 tq -- rounding up is
         * sometimes the closer of the two, and a solver that only ever rounds
         * down silently returns the worse of two available answers.  The
         * cross-check against tools/can_timing.py's exhaustive search is what
         * caught this; it was costing about 10 ppm on inexact rates. */
        for (int round_up = 0; round_up <= 1; round_up++) {
            const uint32_t tq_per_bit = floor_tq + (uint32_t)round_up;
            if (tq_per_bit < 4 || tq_per_bit > 1 + tseg1_max + tseg2_max) {
                continue;
            }
            if (round_up && (clock_hz % (brp * bitrate_bps)) == 0) {
                continue; /* exact: the second rounding is the same candidate */
            }

            /* Sample point splits the bit: sync (1 tq) + tseg1, then tseg2. */
            uint32_t tseg1 =
                (uint32_t)(((uint64_t)tq_per_bit * sample_point_permil) /
                           1000) -
                1;
            if (tseg1 < 1) {
                tseg1 = 1;
            }
            if (tseg1 > tseg1_max) {
                tseg1 = tseg1_max;
            }

            /* Try the ideal split and its neighbours: rounding can put the true
             * best sample point one tq either side. */
            for (int delta = -1; delta <= 1; delta++) {
                const int64_t t1 = (int64_t)tseg1 + delta;
                if (t1 < 1 || t1 > (int64_t)tseg1_max) {
                    continue;
                }
                const int64_t t2 = (int64_t)tq_per_bit - 1 - t1;
                if (t2 < 1 || t2 > (int64_t)tseg2_max) {
                    continue;
                }

                can_timing_t cand = {0};
                cand.brp = brp;
                cand.tseg1 = (uint32_t)t1;
                cand.tseg2 = (uint32_t)t2;
                /* SJW can never exceed phase segment 2, and there is no benefit
                 * to making it smaller than it can be: a larger SJW directly
                 * buys oscillator tolerance. */
                cand.sjw = min_u32((uint32_t)t2, sjw_max);
                cand.nominal_bt_tq = 1 + cand.tseg1 + cand.tseg2;
                cand.bitrate_bps = clock_hz / (brp * cand.nominal_bt_tq);
                cand.sample_point_permil =
                    (int)(((uint64_t)(1 + cand.tseg1) * 1000) /
                          cand.nominal_bt_tq);
                cand.valid = true;

                const int64_t err =
                    ((int64_t)cand.bitrate_bps - (int64_t)bitrate_bps) *
                    1000000 / (int64_t)bitrate_bps;
                cand.bitrate_error_ppm = (int)err;
                cand.max_tolerance_ppm = can_timing_max_tolerance_ppm(&cand);

                const int rate_err = abs_i(cand.bitrate_error_ppm);
                const int sp_err =
                    abs_i(cand.sample_point_permil - sample_point_permil);
                const uint32_t res = min_tq(cand.nominal_bt_tq);

                /* Ranking, most important first.  tools/can_timing.py
                 * implements the same order and tools/tests cross-checks the
                 * two, so this comment is the shared specification rather than
                 * a description of one implementation.
                 *
                 *   1. bit rate error -- a rate that is merely close is a bus
                 *      that works until it does not;
                 *   2. resolution, up to RESOLUTION_ENOUGH_TQ and no further.
                 *      Five time quanta can express 80.0% exactly and still be
                 *      useless: SJW of 1, and no room for transmitter delay
                 *      compensation. Above the floor, extra tq are free and
                 *      should not outrank anything;
                 *   3. sample point, exactly as asked. An earlier sample
                 *      point changes how the bus behaves on a long cable, so
                 *      it is not somewhere to approximate. An earlier version
                 *      of this solver treated misses under 5% as ties and
                 *      quietly
                 *      returned 82.7% for a request of 87.5%, because the wider
                 *      phase segment 2 scored better on tolerance. That is a
                 *      trade for a person to make by asking for a different
                 *      sample point, not one for a solver to make unasked;
                 *   4. oscillator tolerance, among genuine equals;
                 *   5. more tq, purely so the answer is deterministic. */
                bool better;
                if (!best.valid) {
                    better = true;
                } else if (rate_err != best_rate_err) {
                    better = rate_err < best_rate_err;
                } else if (res != best_res) {
                    better = res > best_res;
                } else if (sp_err != best_sp_err) {
                    better = sp_err < best_sp_err;
                } else if (cand.max_tolerance_ppm != best_tol) {
                    better = cand.max_tolerance_ppm > best_tol;
                } else {
                    better = cand.nominal_bt_tq > best.nominal_bt_tq;
                }

                if (better) {
                    best = cand;
                    best_rate_err = rate_err;
                    best_sp_err = sp_err;
                    best_res = res;
                    best_tol = cand.max_tolerance_ppm;
                }
            }
        } /* round_up */
    }

    return best;
}

uint32_t can_timing_nbtp(const can_timing_t *t) {
    if (t == NULL || !t->valid) {
        return 0;
    }
    /* RM0440: NBTP = NSJW[31:25] | NBRP[24:16] | NTSEG1[15:8] | NTSEG2[6:0],
     * every field stored as value-1. */
    uint32_t reg = 0;
    reg |= ((t->sjw - 1) & 0x7Fu) << 25;
    reg |= ((t->brp - 1) & 0x1FFu) << 16;
    reg |= ((t->tseg1 - 1) & 0xFFu) << 8;
    reg |= ((t->tseg2 - 1) & 0x7Fu);
    return reg;
}

uint32_t can_timing_dbtp(const can_timing_t *t, bool tdc_enable) {
    if (t == NULL || !t->valid) {
        return 0;
    }
    /* RM0440: DBTP = DBRP[20:16] | DTSEG1[12:8] | DTSEG2[7:4] | DSJW[3:0],
     * plus TDC at bit 23. */
    uint32_t reg = 0;
    reg |= ((t->brp - 1) & 0x1Fu) << 16;
    reg |= ((t->tseg1 - 1) & 0x1Fu) << 8;
    reg |= ((t->tseg2 - 1) & 0xFu) << 4;
    reg |= ((t->sjw - 1) & 0xFu);
    if (tdc_enable) {
        reg |= (1u << 23);
    }
    return reg;
}

uint32_t can_timing_tdco(const can_timing_t *data_timing,
                         uint32_t transceiver_loop_ns, uint32_t clock_hz) {
    if (data_timing == NULL || !data_timing->valid || clock_hz == 0) {
        return 0;
    }
    /* Place the secondary sample point at the transceiver's loop delay plus
     * the data phase's own sample point, expressed in mtq (clock cycles before
     * the prescaler). */
    const uint64_t loop_cycles =
        ((uint64_t)transceiver_loop_ns * clock_hz) / 1000000000ull;
    const uint64_t offset =
        loop_cycles + (uint64_t)data_timing->brp * (1 + data_timing->tseg1);
    return (uint32_t)(offset > 127 ? 127 : offset);
}
